keps/sig-node/2727-grpc-probe/README.md

```diff
@@ -0,0 +1,177 @@
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture
+      and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in
+  [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents,
+  links to mailing list discussions/SIG meetings, relevant PRs/issues,
+  release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Goals
+
+Enable gRPC probe natively from Kubelet without requiring users to package a
+gRPC healthcheck binary with their container.
+
+- https://github.com/grpc-ecosystem/grpc-health-probe
+- https://github.com/grpc/grpc/blob/master/doc/health-checking.md
+
+## Non-Goals
+
+Add gRPC support in other areas of K8s (e.g. Services).
+
+## Proposal
+
+Add the follow configuration to the `LivenessProbe`, `ReadinessProbe`
+and `StartupProbe`. Example:
+
+```yaml
+    readinessProbe:
+      grpc:                     #+
+        port: 9090              #+
+        service: my-service     #+
+      initialDelaySeconds: 5
+      periodSeconds: 10
+```
+
+This will result in the use of gRPC (using HTTP/2) to use the standard
+healthcheck service to determine the health of the container. We do
+not support other protocols for the time being (HTTP, QUIC), those can
+be enabled in Beta with user feedback.
+
+The healthcheck request will be identified with the following gRPC
+`User-Agent` metadata. This user agent will be statically defined (not
+configurable by the user):
+
+```
+User-Agent: kubernetes/K8S_MAJOR_VER.K8S_MINOR_VER
+```
+
+Example:
+
+```
+User-Agent: kubernetes/1.22
+```
+
+### Risks and Mitigations
+
+1. Adds more code to Kubelet and surface area to Pod.Spec. *Response*: we
+   expect that this will be generally useful given broad gRPC adoption in the
+   industry.
+
+## Design Details
+
+```go
+// core/v1/types.go
+
+type Handler struct {
+  // ...
+  TCPSocket *TCPSocketAction `json...`
+
+  // GRPC specifies an action involving a TCP port.          //+
+  // +optional                                               //+
+  GRPC *GRPCAction `json...`                                 //+
+
+  // ...
+}
+
+type GRPCAction struct {
+  // Port number of the gRPC service. Number must be in the range 1 to 65535.
+  Port int `json:"port" protobuf:"bytes,2,opt,name=port"`
+
+  // Service is the name of the service to place in the gRPC HealthCheckRequest
+  // (see https://github.com/grpc/grpc/blob/master/doc/health-checking.md).
+  //
+  // The service name can be the empty string (i.e. "").
+  Service string `json:"service" protobuf:"bytes,3,opt,name=service"`
```

Comment:
I'm not much into grpc so maybe this is obvious,
can this generate confusion with a kubernetes "service"?

Reply:
i think so

Reply:
added alternatives -- planning to take up during API review